Sources/Foundation/JSONEncoder.swift

```diff
@@ -1152,1359 +1152,801 @@ open class JSONDecoder {
-        self.containers.append(container)
+        case .custom(let closure):
+            return try closure(self)
+        }
     }
+    
+    private func unwrapURL() throws -> URL {
+        let container = SingleValueContainter(impl: self, codingPath: self.codingPath, json: self.json)
+        let string = try container.decode(String.self)
 
-    fileprivate mutating func popContainer() {
-        precondition(!self.containers.isEmpty, "Empty container stack.")
-        self.containers.removeLast()
+        guard let url = URL(string: string) else {
+            throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: self.codingPath,
+                                                                    debugDescription: "Invalid URL string."))
+        }
+        return url
     }
-}
-
-// MARK: Decoding Containers
-
-fileprivate struct _JSONKeyedDecodingContainer<K : CodingKey> : KeyedDecodingContainerProtocol {
-    typealias Key = K
-
-    // MARK: Properties
-
-    /// A reference to the decoder we're reading from.
-    private let decoder: _JSONDecoder
-
-    /// A reference to the container we're reading from.
-    private let container: [String : Any]
-
-    /// The path of coding keys taken to get to this point in decoding.
-    private(set) public var codingPath: [CodingKey]
-
-    // MARK: - Initialization
-
-    /// Initializes `self` by referencing the given decoder and container.
-    fileprivate init(referencing decoder: _JSONDecoder, wrapping container: [String : Any]) {
-        self.decoder = decoder
-        switch decoder.options.keyDecodingStrategy {
-        case .useDefaultKeys:
-            self.container = container
-        case .convertFromSnakeCase:
-            // Convert the snake case keys in the container to camel case.
-            // If we hit a duplicate key after conversion, then we'll use the first one we saw. Effectively an undefined behavior with JSON dictionaries.
-            self.container = Dictionary(container.map {
-                key, value in (JSONDecoder.KeyDecodingStrategy._convertFromSnakeCase(key), value)
-            }, uniquingKeysWith: { (first, _) in first })
-        case .custom(let converter):
-            self.container = Dictionary(container.map {
-                key, value in (converter(decoder.codingPath + [_JSONKey(stringValue: key, intValue: nil)]).stringValue, value)
-            }, uniquingKeysWith: { (first, _) in first })
+    
+    private func unwrapDecimal() throws -> Decimal {
+        guard case .number(let numberString) = self.json else {
+            throw DecodingError.typeMismatch(Decimal.self, DecodingError.Context(codingPath: self.codingPath, debugDescription: ""))
+        }
+        
+        if let decimal = Decimal(string: numberString) {
+            return decimal
         }
-        self.codingPath = decoder.codingPath
+        
+        if let double = Double(numberString) {
+            return Decimal(double)
+        }
```

Comment:
It might be best avoid the fallback to `Double` parsing. Converting a `Double` to `Decimal` is lossy and if the string does not parse as a `Decimal` anyway it can be considered an invalid value for the type.

Reply:
fixed.